Incorrect inspection of a yaml plugin manager

**Environment:**
 - OS: Linux
 - PHP Version: 7.4.8
 - Drupal core: 8.9.6

**Describe the bug**
With a YAML plugin manager, the `__construct` method has to be overridden, but running `drupal-check` produces an error even though the `__construct` exists in the class and does not call the parent `__construct`. 

```
composer create-project drupal/recommended-project:^8.9.2 drupal8
cd drupal8
composer require mglaman/drupal-check phpstan/phpstan:0.12.42 drupal/console drush/drush
vendor/bin/drush si standard --db-url=mysql://[dbuser]:[dbpass]@[location]/[database] -y
vendor/bin/drupal gm --module=test_module --machine-name=test_module -n
vendor/bin/drupal generate:plugin:type:yaml --module=test_module --class=PluginManager -n
vendor/bin/drupal-check web/modules/custom
```

When I looked though the error I can see that https://github.com/mglaman/phpstan-drupal/blob/master/src/Rules/Classes/PluginManagerInspectionRule.php#L109  has `PhpParser\Node\Name` as the `$constructor->class` which doesn't match the class being inspected `Drupal\test_module\PluginManagerManager`

**Console output**

```
$ vendor/bin/drupal-check web/modules/custom/ -vvv
Current working directory: /var/www/drupal8
Using Drupal root: /var/www/drupal8/web
Using vendor root: /var/www/drupal8/vendor
PHPStan configuration:
parameters:
	tipsOfTheDay: false
	reportUnmatchedIgnoredErrors: false
	excludes_analyse:
		- */tests/Drupal/Tests/Listeners/Legacy/*
		- */tests/fixtures/*.php
		- */settings*.php
		- */node_modules/*

	drupal:
		drupal_root: /var/www/drupal8/web

	customRulesetUsed: true
	ignoreErrors:
		- "#\\Drupal calls should be avoided in classes, use dependency injection instead#"
		- "#Plugin definitions cannot be altered.#"
		- "#Missing cache backend declaration for performance.#"
		- "#Plugin manager has cache backend specified but does not declare cache tags.#"

	bootstrapFiles:
		- /var/www/drupal8/vendor/mglaman/drupal-check/src/Command/../../error-bootstrap.php

includes:
	- /var/www/drupal8/vendor/mglaman/drupal-check/src/Command/../../../../phpstan/phpstan-deprecation-rules/rules.neon
	- /var/www/drupal8/vendor/mglaman/drupal-check/src/Command/../../../../mglaman/phpstan-drupal/extension.neon

Result cache not used because the metadata do not match.
 2/2 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100% < 1 sec/< 1 sec 76.5 MiB

 ------ ----------------------------------------------------------------------- 
  Line   test_module/src/PluginManagerManager.php                               
 ------ ----------------------------------------------------------------------- 
  15     Drupal\test_module\PluginManagerManager must override __construct if   
         using YAML plugins.                                                    
  31      cache tag might be unclear and does not contain the cache key in it.  
 ------ ----------------------------------------------------------------------- 

 [ERROR] Found 2 errors
```

--- a/extension.neon
+++ b/extension.neon
@@ -39,7 +39,6 @@ parametersSchema:
 		))
 	])
 rules:
-	- mglaman\PHPStanDrupal\Rules\Classes\PluginManagerInspectionRule
 	- mglaman\PHPStanDrupal\Rules\Drupal\Coder\DiscouragedFunctionsRule
 	- mglaman\PHPStanDrupal\Rules\Drupal\GlobalDrupalDependencyInjectionRule
 	- mglaman\PHPStanDrupal\Rules\Drupal\PluginManager\PluginManagerSetsCacheBackendRule
@@ -81,6 +80,11 @@ services:
 		tags: [phpstan.broker.propertiesClassReflectionExtension]
 	-
 		class: mglaman\PHPStanDrupal\Rules\Classes\ClassExtendsInternalClassRule
+		tags: [phpstan.rules.rule]
+		arguments:
+			reflectionProvider: @reflectionProvider
+	-
+		class: mglaman\PHPStanDrupal\Rules\Classes\PluginManagerInspectionRule
 		tags: [phpstan.rules.rule]
 		arguments:
 			reflectionProvider: @reflectionProvider

--- a/src/Rules/Classes/PluginManagerInspectionRule.php
+++ b/src/Rules/Classes/PluginManagerInspectionRule.php
@@ -4,11 +4,23 @@ namespace mglaman\PHPStanDrupal\Rules\Classes;
 
 use PhpParser\Node;
 use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ReflectionProvider;
 use PHPStan\Rules\Rule;
 use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\ObjectType;
 
+/**
+ * @implements \PHPStan\Rules\Rule<\PhpParser\Node\Stmt\Class_>
+ */
 class PluginManagerInspectionRule implements Rule
 {
+    /** @var ReflectionProvider */
+    private $reflectionProvider;
+    public function __construct(ReflectionProvider $reflectionProvider)
+    {
+        $this->reflectionProvider = $reflectionProvider;
+    }
+
     public function getNodeType(): string
     {
         return Node\Stmt\Class_::class;
@@ -16,15 +28,17 @@ class PluginManagerInspectionRule implements Rule
 
     public function processNode(Node $node, Scope $scope): array
     {
-        assert($node instanceof Node\Stmt\Class_);
-
+        if ($node->namespacedName === null) {
+            // anonymous class
+            return [];
+        }
         if ($node->extends === null) {
             return [];
         }
-
-        // If the class does not extend the default plugin manager, skip it.
-        // @todo inspect interfaces and see if it implements PluginManagerInterface.
-        if ($node->extends->toString() !== 'Drupal\Core\Plugin\DefaultPluginManager') {
+        $className = (string) $node->namespacedName;
+        $pluginManagerType = new ObjectType($className);
+        $pluginManagerInterfaceType = new ObjectType('\Drupal\Component\Plugin\PluginManagerInterface');
+        if (!$pluginManagerInterfaceType->isSuperTypeOf($pluginManagerType)->yes()) {
             return [];
         }
 
@@ -98,15 +112,11 @@ class PluginManagerInspectionRule implements Rule
     {
         $errors = [];
 
-        $fqn = $class->namespacedName;
-        $reflection = new \ReflectionClass($fqn->toString());
+        $fqn = (string) $class->namespacedName;
+        $reflection = $this->reflectionProvider->getClass($fqn);
         $constructor = $reflection->getConstructor();
 
-        if ($constructor === null) {
-            throw new ShouldNotHappenException();
-        }
-
-        if ($constructor->class !== $fqn->toString()) {
+        if ($constructor->getDeclaringClass()->getName() !== $fqn) {
             $errors[] = sprintf('%s must override __construct if using YAML plugins.', $fqn);
         } else {
             foreach ($class->stmts as $stmt) {

--- a/src/Rules/Classes/PluginManagerInspectionRule.php
+++ b/src/Rules/Classes/PluginManagerInspectionRule.php
@@ -99,7 +99,7 @@ class PluginManagerInspectionRule implements Rule
         $errors = [];
 
         $fqn = $class->namespacedName;
-        $reflection = new \ReflectionClass($fqn);
+        $reflection = new \ReflectionClass($fqn->toString());
         $constructor = $reflection->getConstructor();
 
         if ($constructor === null) {

--- a/tests/fixtures/drupal/modules/phpstan_fixtures/src/ExamplePluginManager.php
+++ b/tests/fixtures/drupal/modules/phpstan_fixtures/src/ExamplePluginManager.php
@@ -1,0 +1,105 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\phpstan_fixtures;
+
+use Drupal\Component\Plugin\Exception\PluginException;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Extension\ThemeHandlerInterface;
+use Drupal\Core\Plugin\DefaultPluginManager;
+use Drupal\Core\Plugin\Discovery\ContainerDerivativeDiscoveryDecorator;
+use Drupal\Core\Plugin\Discovery\YamlDiscovery;
+
+/**
+ * Provides the default example manager.
+ */
+class ExamplePluginManager extends DefaultPluginManager {
+
+    /**
+     * The theme handler.
+     *
+     * @var \Drupal\Core\Extension\ThemeHandlerInterface
+     */
+    protected $themeHandler;
+
+    /**
+     * Provides default values for all style_plugin plugins.
+     *
+     * @var array
+     */
+    protected $defaults = [
+        // Add required and optional plugin properties.
+        'id' => '',
+        'enabled' => TRUE,
+        'label' => '',
+        'description' => '',
+        'render' => [],
+    ];
+
+    /**
+     * Constructor.
+     *
+     * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+     *   The module handler.
+     * @param \Drupal\Core\Extension\ThemeHandlerInterface $theme_handler
+     *   The theme handler.
+     * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+     *   Cache backend instance to use.
+     */
+    public function __construct(
+        ModuleHandlerInterface $module_handler,
+        ThemeHandlerInterface $theme_handler,
+        CacheBackendInterface $cache_backend
+    ) {
+        $this->moduleHandler = $module_handler;
+        $this->themeHandler = $theme_handler;
+        $this->alterInfo('ui_examples_examples');
+        $this->setCacheBackend($cache_backend, 'ui_examples', ['ui_examples']);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getDiscovery() {
+        $this->discovery = new YamlDiscovery('ui_examples', $this->moduleHandler->getModuleDirectories() + $this->themeHandler->getThemeDirectories());
+        $this->discovery->addTranslatableProperty('label', 'label_context');
+        $this->discovery->addTranslatableProperty('description', 'description_context');
+        $this->discovery = new ContainerDerivativeDiscoveryDecorator($this->discovery);
+        return $this->discovery;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function processDefinition(&$definition, $plugin_id) : void {
+        parent::processDefinition($definition, $plugin_id);
+        // @todo Add validation of the plugin definition here.
+        if (empty($definition['id'])) {
+            throw new PluginException(sprintf('Example plugin property (%s) definition "id" is required.', $plugin_id));
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function alterDefinitions(&$definitions) {
+        foreach ($definitions as $definition_key => $definition_info) {
+            if (isset($definition_info['enabled']) && !$definition_info['enabled']) {
+                unset($definitions[$definition_key]);
+                continue;
+            }
+        }
+
+        parent::alterDefinitions($definitions);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function providerExists($provider) : bool {
+        return $this->moduleHandler->moduleExists($provider) || $this->themeHandler->themeExists($provider);
+    }
+
+}

--- a/tests/src/Rules/PluginManagerInspectionRuleTest.php
+++ b/tests/src/Rules/PluginManagerInspectionRuleTest.php
@@ -1,0 +1,38 @@
+<?php declare(strict_types=1);
+
+namespace mglaman\PHPStanDrupal\Tests\Rules;
+
+use mglaman\PHPStanDrupal\Rules\Classes\PluginManagerInspectionRule;
+use mglaman\PHPStanDrupal\Tests\DrupalRuleTestCase;
+
+final class PluginManagerInspectionRuleTest extends DrupalRuleTestCase
+{
+
+    protected function getRule(): \PHPStan\Rules\Rule
+    {
+        return new PluginManagerInspectionRule(
+            $this->createReflectionProvider()
+        );
+    }
+
+    /**
+     * @dataProvider pluginManagerData
+     */
+    public function testRule(string $path, array $errorMessages): void
+    {
+        $this->analyse([$path], $errorMessages);
+    }
+
+    public function pluginManagerData(): \Generator
+    {
+        yield 'BreakpointManager' => [
+            __DIR__ . '/../../fixtures/drupal/core/modules/breakpoint/src/BreakpointManager.php',
+            []
+        ];
+        yield 'ExamplePluginManager' => [
+            __DIR__ . '/../../fixtures/drupal/modules/phpstan_fixtures/src/ExamplePluginManager.php',
+            []
+        ];
+    }
+
+}

